js/packages/openinference-instrumentation-langchain/src/tracer.ts

```diff
@@ -0,0 +1,107 @@
+import { BaseTracer, Run } from "@langchain/core/tracers/base";
+import {
+  Tracer,
+  SpanKind,
+  Span,
+  context,
+  trace,
+  SpanStatusCode,
+} from "@opentelemetry/api";
+import { isTracingSuppressed } from "@opentelemetry/core";
+import { SemanticConventions } from "@arizeai/openinference-semantic-conventions";
+import {
+  safelyFlattenAttributes,
+  safelyFormatIO,
+  safelyFormatInputMessages,
+  safelyFormatOutputMessages,
+  safelyGetOpenInferenceSpanKindFromRunType,
+} from "./utils";
+
+type RunWithSpan = {
+  run: Run;
+  span: Span;
+};
+
+export class LangChainTracer extends BaseTracer {
+  private tracer: Tracer;
+  private runs: Record<string, RunWithSpan | undefined> = {};
```

Comment:
This is not getting GC'd

Reply:
the runs? it should when the class gets gc'd which i think is what we want, I'll have to look at the logic a bit more to make sure one of these doesn't get created then infinitely grows but it should be the case that this only gets created when a callback manager gets created and then gets gc'd when the handlers are not used anymore

Reply:
No but you need to evict stuff from this map, otherwise it's hanging on to the mapping for the lifespan of the tracer. This will grow in memory.

Reply:
Ahh i gotcha, yeah i can evict them when the span ends since a parent should never end before it's child correct?

Reply:
might consider https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap